genesis/engine/couplers/sap_coupler.py

```diff
@@ -1740,3 +2393,402 @@ def compute_delassus(self, i_p):
+                pairs[i_p].batch_idx = i_b
+                pairs[i_p].normal = candidates[i_c].normal
+                pairs[i_p].tangent0 = tangent0
+                pairs[i_p].tangent1 = tangent1
+                pairs[i_p].geom_idx0 = i_e
+                pairs[i_p].geom_idx1 = i_f
+                pairs[i_p].vert_idx1 = candidates[i_c].vert_idx1
+                pairs[i_p].barycentric0 = barycentric0
+                pairs[i_p].barycentric1 = barycentric1
+                pairs[i_p].link_idx = i_l
+                pairs[i_p].contact_pos = centroid
+                sap_info[i_p].k = rigid_k
+                sap_info[i_p].phi0 = rigid_phi0
+                sap_info[i_p].mu = ti.sqrt(
+                    self.fem_solver.elements_i[i_e].friction_mu * self.rigid_solver.geoms_info.coup_friction[i_g]
+                )
+            else:
+                overflow = True
+
+        return overflow
+
+    @ti.func
+    def detection(self, f: ti.i32):
+        overflow = False
+        overflow |= self.coupler.rigid_tri_bvh.query(self.coupler.fem_surface_tet_aabb.aabbs)
+        overflow |= self.compute_candidates(f)
+        overflow |= self.compute_pairs(f)
+        return overflow
+
+    @ti.func
+    def compute_W(self):
+        pairs = ti.static(self.contact_pairs)
+        dt2 = self.sim._substep_dt**2
+        # rigid
+        self.coupler.rigid_solve_contact(
+            self.Jt, self.M_inv_Jt, self.n_contact_pairs[None], self.contact_pairs.batch_idx
+        )
+        self.W.fill(0.0)
+        for i_p, i_d, i, j in ti.ndrange(self.n_contact_pairs[None], self.rigid_solver.n_dofs, 3, 3):
+            self.W[i_p][i, j] += self.M_inv_Jt[i_p, i_d][i] * self.Jt[i_p, i_d][j] * dt2
+
+        # fem
+        for i_p in range(self.n_contact_pairs[None]):
+            i_g0 = pairs[i_p].geom_idx0
+            i_b = pairs[i_p].batch_idx
+            for i in ti.static(range(4)):
+                i_v = self.fem_solver.elements_i[i_g0].el2v[i]
+                self.W[i_p] += pairs[i_p].barycentric0[i] ** 2 * self.fem_solver.pcg_state_v[i_b, i_v].prec
+
+    @ti.func
+    def compute_delassus(self, i_p):
+        pairs = ti.static(self.contact_pairs)
+        world = ti.Matrix.cols([pairs[i_p].tangent0, pairs[i_p].tangent1, pairs[i_p].normal])
+        return world.transpose() @ self.W[i_p] @ world
+
+    @ti.func
+    def compute_Jx(self, i_p, x0, x1):
+        pairs = ti.static(self.contact_pairs)
+        i_b = pairs[i_p].batch_idx
+        i_g0 = pairs[i_p].geom_idx0
+        Jx = ti.Vector.zero(gs.ti_float, 3)
+
+        # fem
+        for i in ti.static(range(4)):
+            i_v = self.fem_solver.elements_i[i_g0].el2v[i]
+            Jx = Jx + pairs[i_p].barycentric0[i] * x0[i_b, i_v]
+
+        # rigid
+        for i in range(self.rigid_solver.n_dofs):
+            Jx = Jx - self.Jt[i_p, i] * x1[i_b, i]
+        Jx = ti.Vector([Jx.dot(pairs[i_p].tangent0), Jx.dot(pairs[i_p].tangent1), Jx.dot(pairs[i_p].normal)])
+        return Jx
```

Comment:
Same. Remove final temporary, and not sure `Jx` is the right variable name before that.